### PR TITLE
source: dir: New Directory Source

### PR DESCRIPTION
**Pull request by sakshamarora1**
Example cli:
```console
dffml list records -sources f=dir \
    -source-foldername dataset/train \
    -source-feature image \
    -source-labels airplane bird frog \
    -log debug
```

Labels image arrays with names of their respective folders the images are present in.

| -- dataset
| -- | -- train
| -- | -- | -- label_1
| -- | -- | -- | -- image_1.png
| -- | -- | -- | -- image_2.png
| -- | -- | -- | -- .....
| -- | -- | -- label_2
| -- | -- | -- | -- image_1.png
| -- | -- | -- | -- image_2.png
| -- | -- | -- | -- .....

TODO :-
- [x] Method to save extracted feature vectors in a different file
- [x] Add support for unlabelled images 

Signed-off-by: sakshamarora1 sakshamarora1001@gmail.com

### CODE REVIEW

File: tests/source/test_dir.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            with chdir(tempdir):
+                for dirname in ["a", "b", "c"]:
+                    dirpath = pathlib.Path(dirname)
+                    if not os.path.exists(dirpath):
+                        dirpath.mkdir()
+                    for image_name in ["x", "y", "z"]:
+                        # This is not saved in png file format so isn;t loaded by configloader
+                        # Need to find a better way.
+                        (dirpath / (image_name + ".png")).write_bytes(
+                            bytearray(blank_image.flatten())
```
Comment (pdxjohnny) [inline tests/source/test_dir.py:30]: Let do what we talked about and add PIL and numpy to tests_require and use Image.new here

File: dffml/source/dir.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+import glob
+import pathlib
+from typing import List
+
+from ..record import Record
+from ..base import config, field
+from .memory import MemorySource
+from .file import FileSource
+from ..util.entrypoint import entrypoint
+from dffml.source.source import BaseSource, BaseSourceContext
```
Comment (pdxjohnny) [inline dffml/source/dir.py:14]:
```suggestion
from .source import BaseSource, BaseSourceContext
```

**Comment by pdxjohnny** [comment]
```
2020-06-22T16:35:09.2198822Z ======================================================================
2020-06-22T16:35:09.2198989Z ERROR: test_dir_unlabelled (tests.source.test_dir.TestDirectorySource)
2020-06-22T16:35:09.2199340Z ----------------------------------------------------------------------
2020-06-22T16:35:09.2199500Z Traceback (most recent call last):
2020-06-22T16:35:09.2200062Z   File "/home/runner/work/dffml/dffml/dffml/util/asynctestcase.py", line 68, in run_it
2020-06-22T16:35:09.2200574Z     result = self.loop.run_until_complete(coro(*args, **kwargs))
2020-06-22T16:35:09.2200864Z   File "/usr/share/miniconda/lib/python3.7/asyncio/base_events.py", line 587, in run_until_complete
2020-06-22T16:35:09.2201056Z     return future.result()
2020-06-22T16:35:09.2201356Z   File "/home/runner/work/dffml/dffml/tests/source/test_dir.py", line 80, in test_dir_unlabelled
2020-06-22T16:35:09.2201556Z     async with self.setUpUnlabelledSource() as source:
2020-06-22T16:35:09.2201791Z   File "/home/runner/work/dffml/dffml/dffml/source/dir.py", line 49, in __aenter__
2020-06-22T16:35:09.2201975Z     await self._open()
2020-06-22T16:35:09.2202188Z   File "/home/runner/work/dffml/dffml/dffml/source/dir.py", line 83, in _open
2020-06-22T16:35:09.2202372Z     await self.load_fd()
2020-06-22T16:35:09.2202592Z   File "/home/runner/work/dffml/dffml/dffml/source/dir.py", line 106, in load_fd
2020-06-22T16:35:09.2202781Z     _, feature_data = await cfgl.load_file(image_filename)
2020-06-22T16:35:09.2203025Z   File "/home/runner/work/dffml/dffml/dffml/configloader/configloader.py", line 135, in load_file
2020-06-22T16:35:09.2203238Z     self.parsers, self.async_exit_stack, filepath, base_dir=base_dir
2020-06-22T16:35:09.2203496Z   File "/home/runner/work/dffml/dffml/dffml/configloader/configloader.py", line 109, in load_file
2020-06-22T16:35:09.2203684Z     config_path, conf_dict = await _get_config(path)
2020-06-22T16:35:09.2203917Z   File "/home/runner/work/dffml/dffml/dffml/configloader/configloader.py", line 91, in _get_config
2020-06-22T16:35:09.2204100Z     parsers, exit_stack, temp_filepath, base_dir=base_dir
2020-06-22T16:35:09.2204345Z   File "/home/runner/work/dffml/dffml/dffml/configloader/configloader.py", line 60, in load_single_file
2020-06-22T16:35:09.2204653Z     loader_cls = cls.load(filetype)
2020-06-22T16:35:09.2204879Z   File "/home/runner/work/dffml/dffml/dffml/util/entrypoint.py", line 174, in load
2020-06-22T16:35:09.2205062Z     f"{loading!r} was not found in: {loaded_names}"
2020-06-22T16:35:09.2205549Z dffml.util.entrypoint.EntrypointNotFound: 'png' was not found in: ['json', 'mnistpng']
```

I'm okay to merge this without the example usage in the docstring, but I've created an issue to track that so we make sure it get's done. However, we do need to make sure we fix this ^. Also, I noticed the following which might need to be updated.

```console
$ git grep mnistpng
.gitignore:*.mnistpng
docs/cli.rst:    > four,image1.mnistpng
docs/cli.rst:    > five,image2.mnistpng
docs/cli.rst:    > three,image3.mnistpng
docs/cli.rst:    > two,image4.mnistpng
docs/cli.rst:                "image": "image1.mnistpng"
docs/cli.rst:                "image": "image2.mnistpng"
docs/cli.rst:                "image": "image3.mnistpngFEEDFACE"
docs/cli.rst:                "image": "image4.mnistpng"
```

Whatever's going on here might be an issue for a seperate PR.

**Comment by sakshamarora1** [comment]
@pdxjohnny I think it might be because the new official version hasn't been released yet https://pypi.org/project/dffml-config-png/#history
I uninstalled the development version and re-installed with `python3.7 -m pip install dffml-config-png` and it gave the same `EntrypointNotFound` error

**Pull request merged**